pkg/retriever/prioritywaitqueue/prioritywaitqueue.go

```diff
@@ -0,0 +1,94 @@
+package prioritywaitqueue
+
+import (
+	"fmt"
+	"sync"
+)
+
+// PriorityWaitQueue is a blocking queue for coordinating goroutines, providing
+// a gating mechanism such that only one goroutine may run at a time, where the
+// goroutine allowed to run is chosen based on a priority comparison function.
+type PriorityWaitQueue[T interface{}] interface {
```

Comment:
it might look like overkill but this was so handy for writing tests; my tests all use `int` for this so don't need to get tied up with `QueryResponse`